lib/Db/TagMapper.php

```diff
@@ -0,0 +1,188 @@
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->getTableName())
+			->where(
+					$qb->expr()->eq('id', $qb->createNamedParameter($id, IQueryBuilder::PARAM_INT)),
+					$qb->expr()->eq('user_id', $qb->createNamedParameter($userId))
+				);
+		$result = $this->findEntity($qb);
+		return $result;
+	}
+
+	/**
+	 * Tag a message in the DB
+	 *
+	 * To tag (flag) a message on IMAP, @see OCA\Mail\Service\MailManager::tagMessage
+	 *
+	 * @param Tag $tag
+	 * @param string $messageId
+	 * @param string $userId
+	 * @return void
+	 */
+	public function tagMessage(Tag $tag, string $messageId, string $userId) {
+		/** @var Tag $exists */
+		try {
+			$exists = $this->getTagByImapLabel($tag->getImapLabel(), $userId);
+			$tag->setId($exists->getId());
+		} catch (DoesNotExistException $e) {
+			$tag = $this->insert($tag);
+		}
+
+		$qb = $this->db->getQueryBuilder();
+		$qb->insert('mail_message_tags');
+		$qb->setValue('imap_message_id', $qb->createNamedParameter($messageId));
+		$qb->setValue('tag_id', $qb->createNamedParameter($tag->getId(), IQueryBuilder::PARAM_INT));
+		$qb->execute();
+	}
+
+	/**
+	 * Remove a tag from a DB message
+	 *
+	 * @param Tag $tag
+	 * @param string $messageId
+	 * @return void
+	 */
+	public function untagMessage(Tag $tag, string $messageId) {
+		$qb = $this->db->getQueryBuilder();
+		$qb->delete('mail_message_tags')
+			->where($qb->expr()->eq('imap_message_id', $qb->createNamedParameter($messageId)))
+			->where($qb->expr()->eq('tag_id', $qb->createNamedParameter($tag->getId())));
+		$qb->execute();
+	}
+
+	/**
+	 * Create some default system tags
+	 *
+	 * This is designed after Thunderbird's email tags
+	 * $label1 to $label5 with the according states and colours
+	 *
+	 * @param MailAccount $account
+	 * @return void
+	 *
+	 * @link https://github.com/nextcloud/mail/issues/25
+	 */
+	public function createDefaultTags(MailAccount $account) {
+		for ($i = 1; $i < 6; $i++) {
+			$tag = new \OCA\Mail\Db\Tag();
+			$tag->setImapLabel('$label' . $i);
+			$tag->setUserId($account->getUserId());
+			switch ($i) {
+				case $i === 1:
+					$tag->setDisplayName('Important');
+					$tag->setColor('#FF0000');
+					$tag->setIsSystemFlag(1);
+
+					break;
+				case $i === 2:
+					$tag->setDisplayName('Work');
+					$tag->setColor('#FFC300');
+					$tag->setIsSystemFlag(1);
+
+					break;
+				case $i === 3:
+					$tag->setDisplayName('Personal');
+					$tag->setColor('#008000');
+					$tag->setIsSystemFlag(1);
+
+					break;
+				case $i === 4:
+					$tag->setDisplayName('To Do');
+					$tag->setColor('#000080');
+					$tag->setIsSystemFlag(1);
+
+					break;
+				case $i === 5:
+					$tag->setDisplayName('Later');
+					$tag->setColor('#800080');
+					$tag->setIsSystemFlag(1);
+					break;
+			}
+			try {
+				$this->insert($tag);
+			} catch (\Doctrine\DBAL\Exception\UniqueConstraintViolationException $e) {
+				$this->logger->debug($e->getMessage());
```

Comment:
when logging an exception pass the object to the logger, so the trace is included in the log line.

```suggestion
				$this->logger->debug($e->getMessage(), ['exception' => $e]);
```